File: PadroesEstruturais/Decorator/decorator01.py
# Componente
class ComponenteNotificador:
    def enviar(self, msg):
        print(f"Enviando mensagem: {msg}")

# Decorador Base:
class DecoradorNotificador:
    def __init__(self, componente):
        self.componente = componente
    
    def enviar(self, msg):
        self.componente.enviar(msg)

# Decorador Concreto 1 - Simulando SMS:
class NotificadorSMS(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f"Enviando mensagem SMS: {msg}")

# Decorador Concreto 2 - Simulando WhatsApp:
class NotificadorWhatsApp(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f"Enviando mensagem WhatsApp: {msg}")
    
# Decorador Concreto 3 - Simulando Email:
class NotificadorEmail(DecoradorNotificador):
    def enviar(self, msg):
        super().enviar(msg)
        print(f"Enviando mensagem Email: {msg}")
    
# Interface de Utilização ------------------------------------
notificador = ComponenteNotificador()
notificadorSMS = NotificadorSMS(notificador)
notificadorWhatsApp = NotificadorWhatsApp(notificador)
notificadorEmail = NotificadorEmail(notificador)

mensagem = "Ola Mundo pelo Python... "

notificadorWhatsApp.enviar(mensagem)
notificadorEmail.enviar(mensagem)
notificadorSMS.enviar(mensagem)

File: PadroesEstruturais/Decorator/decorator01.js
// Componente 
class ComponenteNotificador {
    enviar(msg) {
        console.log(`Enviando mensagem: ${msg}`)
    }
}


// Decorator Base
class DecoratorNotificador {
    constructor(componente) {
        this.componente = componente;
    }

    enviar(msg) {
        this.componente.enviar(msg)
    }
    
}


// Decorator Concreto 1 - Simulando SMS:
class NotificadorSMS extends DecoratorNotificador {
    enviar(msg) {
        super.enviar(msg);
        console.log(`Enviando mensagem por SMS: ${msg}`);
    }
}


// Decorator Concreto 2 - Simulando WhatsApp:
class NotificadorWhatsApp extends DecoratorNotificador {
    enviar(msg) {
        super.enviar(msg);
        console.log(`Enviando mensagem por WhatsApp: ${msg}`);
    }
}


// Decorator Concreto 3 - Simulando Email:
class NotificadorEmail extends DecoratorNotificador {
    enviar(msg) {
        super.enviar(msg);
        console.log(`Enviando mensagem por Email: ${msg}`);
    }
}


class NotificadorFacebook extends DecoratorNotificador {
    enviar(msg) {
        super.enviar(msg);
        console.log(`Enviando mensagem por Facebook: ${msg}`);
    }
}


// Interface de Utilização -----------------------------
const notificador = new ComponenteNotificador();
const notificadorSMS = new NotificadorSMS(notificador);
const notificadorWhatsApp = new NotificadorWhatsApp(notificador);
const notificadorEmail = new NotificadorEmail(notificador);
const notificadorFacebook = new NotificadorFacebook(notificador);

const mensagem = "Olá Mundo.."
notificadorWhatsApp.enviar(mensagem);
notificadorEmail.enviar(mensagem);
notificadorSMS.enviar(mensagem);
notificadorFacebook.enviar(mensagem);






